homeassistant/components/yale_smart_alarm/lock.py

```diff
@@ -0,0 +1,135 @@
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import (
+    CONF_LOCK_CODE_DIGITS,
+    COORDINATOR,
+    DEFAULT_LOCK_CODE_DIGITS,
+    DOMAIN,
+    LOGGER,
+    MANUFACTURER,
+    MODEL,
+)
+from .coordinator import YaleDataUpdateCoordinator
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Yale lock entry."""
+
+    coordinator: YaleDataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id][
+        COORDINATOR
+    ]
+    code_format = entry.options.get(CONF_LOCK_CODE_DIGITS, DEFAULT_LOCK_CODE_DIGITS)
+
+    async_add_entities(
+        YaleDoorlock(coordinator, data, code_format)
+        for data in coordinator.data["locks"]
+    )
+
+
+class YaleDoorlock(CoordinatorEntity, LockEntity):
+    """Representation of a Yale doorlock."""
+
+    def __init__(
+        self, coordinator: YaleDataUpdateCoordinator, data: dict, code_format: int
+    ) -> None:
+        """Initialize the Yale Lock Device."""
+        super().__init__(coordinator)
+        self._coordinator = coordinator
+        self._attr_name = data["name"]
+        self._attr_unique_id = data["address"]
+        self._attr_device_info = DeviceInfo(
+            name=self._attr_name,
+            manufacturer=MANUFACTURER,
+            model=MODEL,
+            identifiers={(DOMAIN, data["address"])},
+            via_device=(DOMAIN, self._coordinator.entry.data[CONF_USERNAME]),
+        )
+        self._attr_code_format = f"^\\d{code_format}$"
+
+    async def async_unlock(self, **kwargs) -> None:
+        """Send unlock command."""
+        code = kwargs.get(ATTR_CODE, self._coordinator.entry.options.get(CONF_CODE))
+
+        if not code or not self._coordinator.yale:
+            return
+
+        try:
+            get_lock = await self.hass.async_add_executor_job(
+                self._coordinator.yale.lock_api.get, self._attr_name
+            )
+            lock_state = await self.hass.async_add_executor_job(
+                self._coordinator.yale.lock_api.open_lock,
+                get_lock,
+                code,
+            )
+        except (
+            AuthenticationError,
+            ConnectionError,
+            TimeoutError,
+            UnknownError,
+        ) as error:
+            raise HomeAssistantError(
+                f"Could not verify unlocking for {self._attr_name}: {error}"
+            ) from error
+
+        LOGGER.debug("Door unlock: %s", lock_state)
+        if lock_state:
+            self._attr_is_locked = False
+            self.async_write_ha_state()
+            return
+        raise HomeAssistantError("Could not unlock, check system ready for unlocking")
+
+    async def async_lock(self, **kwargs) -> None:
+        """Send lock command."""
+
+        if not self._coordinator.yale:
```

Comment:
Should this raise?

Reply:
It's only there to make mypy happy. If yale could not load in the coordinator then the coordinator would raise errors during loading of the integration and this check is of course then pointless.

Reply:
Got another recommendation from Martin on separate PR so this now checks typing and asserts